daml-lf/transaction/src/test/scala/com/digitalasset/daml/lf/transaction/TransactionNodesStatisticsSpec.scala

```diff
@@ -0,0 +1,173 @@
+      observers = Set.empty,
+      key = if (withKey) Some(Value.ValueUnit) else None,
+      maintainers = if (withKey) parties else Set.empty,
+      byInterface = None,
+    )
+  }
+
+  private[this] def exe(consuming: Boolean, byKey: Boolean)(b: TxBuilder) = {
+    val c = create(b, byKey)
+    b.exercise(
+      contract = c,
+      choice = b.newChoiceName,
+      consuming = consuming,
+      actingParties = c.signatories,
+      argument = Value.ValueUnit,
+      result = Some(Value.ValueUnit),
+      choiceObservers = Set.empty,
+      byKey = byKey,
+      byInterface = None,
+    )
+  }
+
+  private[this] def fetch(byKey: Boolean)(b: TxBuilder) =
+    b.fetch(create(b, byKey), byKey, None)
+
+  private[this] def lookup(b: TxBuilder) =
+    b.lookupByKey(create(b, withKey = true), true)
+
+  private[this] def rollback(b: TxBuilder) =
+    b.rollback()
+
+  private[this] def addAllNodes(b: TxBuilder, rbId: NodeId) = {
+    b.add(create(b), rbId)
+    b.add(exe(consuming = false, byKey = false)(b), rbId)
+    b.add(exe(consuming = true, byKey = false)(b), rbId)
+    b.add(exe(consuming = false, byKey = true)(b), rbId)
+    b.add(exe(consuming = true, byKey = true)(b), rbId)
+    b.add(fetch(byKey = false)(b), rbId)
+    b.add(fetch(byKey = true)(b), rbId)
+    b.add(lookup(b), rbId)
+    b.add(rollback(b), rbId)
+    ()
+  }
+
+  "TransactionNodeStatistics#+" should {
+
+    "add" in {
+
+      val s1 = TransactionNodesStatistics(1, 1, 1, 1, 1, 1, 1, 1, 1)
+      val s2 = TransactionNodesStatistics(2, 3, 5, 7, 11, 13, 17, 19, 23)
+      val s1s2 = TransactionNodesStatistics(3, 4, 6, 8, 12, 14, 18, 20, 24)
+      val s2s2 = TransactionNodesStatistics(4, 6, 10, 14, 22, 26, 34, 38, 46)
+
+      s2 + TransactionNodesStatistics.Empty shouldBe s2
+      TransactionNodesStatistics.Empty + s2 shouldBe s2
+      s1 + s2 shouldBe s1s2
+      s2 + s1 shouldBe s1s2
+      s2 + s2 shouldBe s2s2
+    }
+
+  }
+
+  "TransactionNodeStatistics.stats" should {
+
+    val n = 3
+
+    type Getter = TransactionNodesStatistics => Int
+    val testCases = Table[TxBuilder => Node, Getter](
+      "makeNode" -> "getter",
+      (create(_), _.creates),
+      (exe(consuming = true, byKey = false), _.consumingExercisesByCid),
+      (exe(consuming = false, byKey = false), _.nonconsumingExercisesByCid),
+      (exe(consuming = true, byKey = true), _.consumingExercisesByKey),
+      (exe(consuming = false, byKey = true), _.nonconsumingExercisesByKey),
+      (fetch(byKey = false), _.fetchesByCid),
+      (fetch(byKey = true), _.fetchesByKey),
+      (lookup, _.lookupsByKey),
+      (rollback, _.rollbacks),
+    )
+
+    "count each type of committed nodes properly" in {
+      forEvery(testCases) { (makeNode, getter) =>
+        val builder = TxBuilder()
+
+        for (i <- 1 to n) {
+          builder.add(makeNode(builder))
+          inside(TransactionNodesStatistics.stats(builder.build())) {
+            case (committed, rollbacked) =>
```

Comment:
```suggestion
            case (committed, rolledBack) =>
```
Same across the file